### PR TITLE
Additional FX Patches

### PR DESCRIPTION
**Pull request by denopink**
What's included:
- [v2 & v3 endpoints are not enforcing capabilities xmidt-org/tr1d1um#342](https://github.com/xmidt-org/tr1d1um/issues/342)
- [v3 endpoint is not validating webhooks #341](https://github.com/xmidt-org/tr1d1um/issues/341)
- [8090 support has been removed #343](https://github.com/xmidt-org/tr1d1um/issues/343)

Dependent on:
- [x] https://github.com/xmidt-org/tr1d1um/pull/324
- [x] https://github.com/xmidt-org/argus/pull/333
- [x] https://github.com/xmidt-org/ancla/pull/132

### CODE REVIEW

File: routes.go [hunk 1]
```diff
@@ -231,13 +247,15 @@ func handlePrimaryEndpoint(in primaryEndpointIn) {
...
 		if err != nil {
 			fmt.Fprintf(os.Stderr, "Failed to initialize v2 endpoint middleware: %v\n", err)
 			return err
 		}
-		in.APIRouter.Handle("/hook", in.AuthChain.Then(fixV2Middleware(in.GetAllWebhooksHandler))).Methods(http.MethodPost)
-		in.APIRouter.Handle("/hooks", in.AuthChain.Then(in.GetAllWebhooksHandler)).Methods(http.MethodGet)
+		in.PrimaryRouter.Handle("/hook", in.AuthChain.Then(fixV2Middleware(in.AddWebhookHandler))).Methods(http.MethodPost)
+		in.PrimaryRouter.Handle("/hooks", in.AuthChain.Then(in.GetAllWebhooksHandler)).Methods(http.MethodGet)
+		in.AltRouter.Handle("/hook", in.AuthChain.Then(fixV2Middleware(in.AddWebhookHandler))).Methods(http.MethodPost)
+		in.PrimaryRouter.Handle("/hooks", in.AuthChain.Then(in.GetAllWebhooksHandler)).Methods(http.MethodGet)
```
[the remark this answers is not in the document]
Comment (denopink) [reply routes.go]: @mtrinh11 fyi

**Review by renaz6 — APPROVED** [review]

File: routes.go [hunk 2]
```diff
@@ -267,8 +273,16 @@ func metricMiddleware(f *touchstone.Factory) (out metricMiddlewareOut) {
...
+	return in.PrimaryRouter.PathPrefix(in.URLPrefix).Subrouter()
+}
+
+func buildAPIAltRouter(in apiAltRouterIn) {
+	apiAltRouter := in.AlternateRouter.PathPrefix(in.URLPrefix).Subrouter()
+	apiAltRouter.Handle("/device/{deviceid}/{service}", in.APIRouter)
+	apiAltRouter.Handle("/device/{deviceid}/{service}/{parameter}", in.APIRouter)
+	apiAltRouter.Handle("/device/{deviceid}/stat", in.APIRouter)
+	apiAltRouter.Handle("/hook", in.APIRouter)
+	apiAltRouter.Handle("/hooks", in.APIRouter)
```
Comment (denopink) [inline routes.go:285]: @johnabass I created this quick patch to get port 8090 back on tr1d1um

File: transaction/errors.go [hunk 3]
```diff
@@ -63,11 +61,9 @@ func NewCodedError(e error, code int) CodedError {
 
 // ErrorLogEncoder decorates the errorEncoder in such a way that
 // errors are logged with their corresponding unique request identifier
-func ErrorLogEncoder(getLogger GetLoggerFunc, ee kithttp.ErrorEncoder) kithttp.ErrorEncoder {
+func ErrorLogEncoder(getLogger sallust.GetLoggerFunc, ee kithttp.ErrorEncoder) kithttp.ErrorEncoder {
 	if getLogger == nil {
```
Comment (johnabass) [inline transaction/errors.go:65]: No need to check for nil.  A nil logger should indicate a bug,  which is what panics are for

File: transaction/errors_test.go [hunk 4]
```diff
@@ -45,15 +46,15 @@ func TestBadRequestError(t *testing.T) {
 func TestErrorLogEncoder(t *testing.T) {
 	tcs := []struct {
 		desc      string
-		getLogger GetLoggerFunc
+		getLogger sallust.GetLoggerFunc
```
Comment (johnabass) [inline transaction/errors_test.go:49]: sallust shouldn't export a function type for this.  It requires apps to express a dependency in many cases just to pull in this type.  Using func(context.Context) *zap.Logger as the type suffices.

**Review by johnabass — CHANGES_REQUESTED** [review]

File: transaction/errors_test.go (hunk 4 above)
Comment (denopink) [reply transaction/errors_test.go:49]: my bad, thought I removed all these references. Double checking other servers as well

File: routes.go [hunk 5]
```diff
@@ -312,10 +322,11 @@ func fixV2Duration(getLogger ancla.GetLoggerFunc, config ancla.TTLVConfig, v2Han
 			// if this is v2, we need to unmarshal and check the duration.  If
 			// the duration is bad, change it to 5m and add a header. Then use
 			// the v2 handler.
-			logger := getLogger(r.Context())
+			logger := sallusthttp.Get(r)
 			if logger == nil {
-				logger = log.NewNopLogger()
+				panic("no logger supplied")
```
Comment (johnabass) [inline routes.go:327]: you don't need to explicitly panic here.  sallust.Get is documented never to return nil.

**Review by johnabass — CHANGES_REQUESTED** [review]

File: primaryHandler.go [hunk 6]
```diff
@@ -167,19 +148,19 @@ func provideWebhookHandlers(in provideWebhookHandlersIn) (out provideWebhookHand
 		return
 	}
 
-	webhookConfig := in.WebhookConfigKey
-	webhookConfig.Logger = gokitLogger(in.Logger)
+	webhookConfig := in.WebhookConfig
+	webhookConfig.Logger = in.Logger
```
Comment (johnabass) [inline primaryHandler.go:152]: Is this just because the `webhookConfig` type hasn't yet been updated to remove `go-kit`'s logger?

**Review by johnabass — CHANGES_REQUESTED** [review]

File: primaryHandler.go (hunk 6 above)
Comment (denopink) [reply primaryHandler.go:152]: Ah, yes this one sneak by me
I'll verify all of these have been accounted for

Comment (denopink) [reply primaryHandler.go:152]: I miss read your comment, if you're referring to `cannot use in.Logger (variable of type *zap.Logger) as "github.com/go-kit/log".Logger value in assignment: *zap.Logger does not implement "github.com/go-kit/log".Logger (wrong type for method Log)` then yes, that's what the https://github.com/xmidt-org/argus/pull/333 take care of

**Pull request merged**